tox.ini

```diff
@@ -21,6 +21,7 @@ deps =
     mock
```

Comment:
A better solution to this is to just switch to isolated build https://tox.readthedocs.io/en/latest/config.html#conf-isolated_build

Reply:
I've tried it at first, but this doesn't fix the issue. I still get the error:

```
distutils.errors.DistutilsError: Could not find suitable distribution for Requirement.parse('setuptools_scm')
```

Reply:
from where how? it works for tox master https://github.com/tox-dev/tox/blob/master/pyproject.toml#L1

Reply:
I've checked it on tox master. See:

```bash
#
# ======== Ran with the internet ========
#
$ tox -e py37 -- --version
.package create: /Users/albert/Projects/tox/.tox/.package
.package installdeps: setuptools >= 40.0.4, setuptools_scm >= 2.0.0, <4, wheel >= 0.29.0
py37 create: /Users/albert/Projects/tox/.tox/py37
py37 installdeps: pip == 19.1.1
py37 inst: /Users/albert/Projects/tox/.tox/.tmp/package/1/tox-3.14.1.dev3+gb0c5331.tar.gz
py37 installed: apipkg==1.5,atomicwrites==1.3.0,attrs==19.2.0,coverage==4.5.4,execnet==1.7.1,filelock==3.0.12,flaky==3.6.1,freezegun==0.3.12,importlib-metadata==0.23,more-itertools==7.2.0,packaging==19.2,pathlib2==2.3.5,pluggy==0.13.0,psutil==5.6.3,py==1.8.0,pyparsing==2.4.2,pytest==5.2.1,pytest-cov==2.8.1,pytest-forked==1.1.1,pytest-mock==1.11.1,pytest-randomly==1.2.3,pytest-xdist==1.30.0,python-dateutil==2.8.0,six==1.12.0,toml==0.10.0,tox==3.14.1.dev3+gb0c5331,virtualenv==16.7.5,wcwidth==0.1.7,zipp==0.6.0
py37 run-test-pre: PYTHONHASHSEED='545282946'
py37 run-test: commands[0] | pytest --cov /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/tox --cov-config /Users/albert/Projects/tox/tox.ini --junitxml /Users/albert/Projects/tox/.tox/junit.py37.xml -n=auto --version
This is pytest version 5.2.1, imported from /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pytest.py
setuptools registered plugins:
  pytest-xdist-1.30.0 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/xdist/plugin.py
  pytest-xdist-1.30.0 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/xdist/looponfail.py
  pytest-randomly-1.2.3 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pytest_randomly.py
  pytest-forked-1.1.1 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pytest_forked/__init__.py
  flaky-3.6.1 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/flaky/flaky_pytest_plugin.py
  pytest-mock-1.11.1 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pytest_mock/__init__.py
  pytest-cov-2.8.1 at /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pytest_cov/plugin.py
__________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
  py37: commands succeeded
  congratulations :)

#
# ======== Ran without the internet ========
#
$ tox -e py37 -- --version
py37 inst-nodeps: /Users/albert/Projects/tox/.tox/.tmp/package/1/tox-3.14.1.dev3+gb0c5331.tar.gz
ERROR: invocation failed (exit code 1), logfile: /Users/albert/Projects/tox/.tox/py37/log/py37-4.log
================================================================================================================= log start ==================================================================================================================
Processing ./.tox/.tmp/package/1/tox-3.14.1.dev3+gb0c5331.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
  ERROR: Complete output from command /Users/albert/Projects/tox/.tox/py37/bin/python /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/4x/rz_w89dx1z3bydfd6qpdjn_h0000gn/T/pip-build-env-4llf8ugr/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools >= 40.0.4' 'setuptools_scm >= 2.0.0, <4' 'wheel >= 0.29.0':
  ERROR: Collecting setuptools>=40.0.4
    WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x109ae9c88>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/setuptools/
    WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x109ae9c18>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/setuptools/
    WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x109ae9a58>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/setuptools/
    WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x109ae99e8>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/setuptools/
    WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x109ae98d0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/setuptools/
    ERROR: Could not find a version that satisfies the requirement setuptools>=40.0.4 (from versions: none)
  ERROR: No matching distribution found for setuptools>=40.0.4
  WARNING: You are using pip version 19.1.1, however version 19.3 is available.
  You should consider upgrading via the 'pip install --upgrade pip' command.
  ----------------------------------------
ERROR: Command "/Users/albert/Projects/tox/.tox/py37/bin/python /Users/albert/Projects/tox/.tox/py37/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/4x/rz_w89dx1z3bydfd6qpdjn_h0000gn/T/pip-build-env-4llf8ugr/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools >= 40.0.4' 'setuptools_scm >= 2.0.0, <4' 'wheel >= 0.29.0'" failed with error code 1 in None
WARNING: You are using pip version 19.1.1, however version 19.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

================================================================================================================== log end ===================================================================================================================
__________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
ERROR:   py37: InvocationError for command /Users/albert/Projects/tox/.tox/py37/bin/python -m pip install --no-deps -U '/Users/albert/Projects/tox/.tox/.tmp/package/1/tox-3.14.1.dev3+gb0c5331.tar.gz[testing]' (exited with code 1)```

Reply:
that seems like a pip bug 🤔 does not hit the pip cache when provisioning build dependencies?

Reply:
It would appear so. I'll file an issue on pip's issue tracker.

Reply:
The problem is in this command:
`pip install --ignore-installed --no-user --prefix /private/var/folders/4x/rz_w89dx1z3bydfd6qpdjn_h0000gn/T/pip-build-env-4llf8ugr/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools >= 40.0.4' 'setuptools_scm >= 2.0.0, <4' 'wheel >= 0.29.0'`

and look like `--ignore-installed` option doesn't honour the cache.

Reply:
🤔 definitely a pip bug 😄

Reply:
I'm inclined to think that's the correct behavior. To install a package (even from a cache) It needs the network to get the info about available versions.

Reply:
Not really. The intent of that definition says give me ``'setuptools >= 40.0.4'``, not that give me the latest available; if that can be satisfied from cache then I'd argue it suffices to keep the network local 🤔

Reply:
maybe pip should have a --try-first-cache flag 🤔

Reply:
Also, I've found that since pip>=18.1, pip always needs the network. Before that, it could install cached without the internet. I'll try to git-bisect it.

Reply:
@gaborbernat I was introduced in https://github.com/pypa/pip/pull/5791

So requests had been caching for 10 minutes (before this PR), and technically pip always required the internet.

Reply:
> that seems like a pip bug 🤔 does not hit the pip cache when provisioning build dependencies?

I wonder what @pradyunsg would think.

Reply:
I'm not sure I follow what the issue here is, that I'm being asked to provide my thoughts on. :)

Reply:
@pradeepcep it would be good if isolated builds didn't need to use the network if the cache has the build deps in already

Reply:
@pradyunsg *